packages/rum/src/parentContexts.ts

```diff
@@ -16,45 +21,132 @@ export interface ActionContext extends Context {
+  let currentAction: CurrentContext | undefined
   let currentSessionId: string | undefined
 
-  lifeCycle.subscribe(LifeCycleEventType.VIEW_CREATED, (internalContext) => {
-    currentView = internalContext
+  let previousViews: Array<PreviousContext<ViewContext>> = []
+  let previousActions: Array<PreviousContext<ActionContext>> = []
+
+  lifeCycle.subscribe(LifeCycleEventType.VIEW_CREATED, (currentContext) => {
+    if (currentView && withContextHistory) {
+      previousViews.unshift({
+        context: buildCurrentViewContext(),
+        endTime: currentContext.startTime,
+        startTime: currentView.startTime,
+      })
+    }
+    currentView = currentContext
     currentSessionId = session.getId()
   })
 
-  lifeCycle.subscribe(LifeCycleEventType.AUTO_ACTION_CREATED, (internalContext) => {
-    currentAction = internalContext
+  lifeCycle.subscribe(LifeCycleEventType.AUTO_ACTION_CREATED, (currentContext) => {
+    currentAction = currentContext
   })
 
-  lifeCycle.subscribe(LifeCycleEventType.AUTO_ACTION_COMPLETED, () => {
+  lifeCycle.subscribe(LifeCycleEventType.AUTO_ACTION_COMPLETED, (userAction: AutoUserAction) => {
+    if (currentAction && withContextHistory) {
+      previousActions.unshift({
+        context: buildCurrentActionContext(),
+        endTime: currentAction.startTime + userAction.duration,
+        startTime: currentAction.startTime,
+      })
+    }
     currentAction = undefined
   })
 
   lifeCycle.subscribe(LifeCycleEventType.AUTO_ACTION_DISCARDED, () => {
     currentAction = undefined
   })
 
+  lifeCycle.subscribe(LifeCycleEventType.SESSION_RENEWED, () => {
+    previousViews = []
+    previousActions = []
+    currentView = undefined
+    currentAction = undefined
+  })
+
+  const clearOldContextsInterval = window.setInterval(
+    monitor(() => {
+      clearOldContexts(previousViews, VIEW_CONTEXT_TIME_OUT_DELAY)
+      clearOldContexts(previousActions, ACTION_CONTEXT_TIME_OUT_DELAY)
+    }),
+    CLEAR_OLD_CONTEXTS_INTERVAL
+  )
+
+  function clearOldContexts(previousContexts: Array<PreviousContext<unknown>>, timeOutDelay: number) {
+    const oldTimeThreshold = performance.now() - timeOutDelay
+    while (previousContexts.length > 0 && previousContexts[previousContexts.length - 1].startTime < oldTimeThreshold) {
+      previousContexts.pop()
+    }
+  }
+
+  function buildCurrentViewContext() {
+    return { sessionId: currentSessionId, view: { id: currentView!.id, url: location.href } }
+  }
+
+  function buildCurrentActionContext() {
+    return { userAction: { id: currentAction!.id } }
+  }
+
+  function findContext<T>(
+    buildContext: () => T,
+    previousContexts: Array<PreviousContext<T>>,
+    currentContext?: CurrentContext,
+    startTime?: number
+  ) {
+    if (!startTime) {
+      return currentContext ? buildContext() : undefined
+    }
+    if (currentContext && startTime >= currentContext.startTime) {
+      return buildContext()
+    }
+    if (!withContextHistory) {
```

Comment:
Maybe should you extract this `if (!withContextHistory)` outside the scope of `findContext` as this function should not be called if `withContextHistory=false`.

I would rather :

```
  findView: (startTime) => {
      return withContextHistory ? findContext(buildCurrentViewContext, previousViews, currentView, startTime) : undefined
    },
```

Also by doing that, you could remove `findContext` from the body of the `startParentContexts`. That would reduce `startParentContexts` size and clarify its inner code.

Reply:
Keep in mind that this `if (!withContextHistory)` will be removed when we will clean the feature flag.
By moving this outside of findContext, we should need two different functions:

1. `findContextWithoutHistory`
2. `findContextWithHistory`

the first one should be called by the second one to avoid logic duplication, is that what you mean?

Reply:
Why would you need two functions ? Isn't possible to call `findContext` only if there is an history. Otherwise, in `findView` and `findAction`, instead of calling `findContext` you return undefined.

When you remove the FF you will just have to clean `findView` and `findAction`.

Reply:
I don't get what you are saying. In `findContext`, there is some logic before `if (!withContextHistory)`, returning undefined when the flag is false would break the current behavior.

Reply:
Yes I mean that this code, before the `findContext` does not really find the context in the history but build one instead (using the function you pass as a parameter).

A solution would have been to handle these case in the findView / findAction without passing a buildContext param and either findContext if there is an history or build one (using buildCurrentActionContext/buildCurrentViewContext) if there is none.

Reply:
oh ok, findContext is not related only to history, like findAction/View.
I found convenient to have the same kind of naming but I am open other naming suggestion if you found it confusing.
However, I found that having the whole strategy at a single place is pretty convenient since it is not that big. IMO, moving some parts of it outside of startParentContext does not seem to bring much value.